pirate/mem.c

```diff
@@ -6,35 +6,172 @@
+    if (s_BigBuf == NULL) {
+        assert(false); // BigBuffer_Initialize() must be called before BigBuffer_calloc()
         return NULL;
     }
+    if ((alignment & (alignment - 1)) != 0u) {
+        assert(false); // alignment must be a power of 2
+        return NULL;
+    }
+    if (elementCount == 0 || elementSize == 0) {
+        assert(false); // calloc is undefined for zero elements or zero size
+        return NULL;
+    }
+    if (alignment == 0) {
+        alignment = 1; // fixup common API error
+    }
+    if (elementCount > (SIZE_MAX / elementSize)) {
+        // calculating total size would overflow
+        return NULL;
+    }
+    size_t totalRequestedBytes = elementCount * elementSize;
+    // exit early if the request number of bytes is larger than what's left
+    // Note: it might still not fit due to alignment requirements ... checked later.
+    if (totalRequestedBytes > (size_t)(s_BigBufHighWaterMark - s_BigBufLowWaterMark)) {
+        return NULL;
+    }
+
+    // Are we allocating from the bottom (large alignment) or from the top (small alignment)?
+    bool allocateFromBottom = (alignment >= 1024) || (totalRequestedBytes >= 1024);
+    uint8_t * result =
+        allocateFromBottom ?
+        s_BigBufLowWaterMark :
+        s_BigBufHighWaterMark - totalRequestedBytes;
+
+    // If result pointer is NOT already aligned, adjust it.
+    size_t alignmentMask = alignment - 1;
+    if (((size_t)result & alignmentMask) != 0) {
+        if (allocateFromBottom) {
+            // Allocation must be moved to larger address
+            result = (uint8_t*)( ((size_t)result + alignment) & ~alignmentMask );
+            if ((size_t)s_BigBufHighWaterMark - (size_t)result < totalRequestedBytes) {
+                // insufficient space to allocate, due to alignment requirement
+                return NULL;
+            }
+        } else {
+            // Allocation must be moved to smaller address
+            result = (uint8_t*)( (size_t)result & ~alignmentMask );
+            if (result < s_BigBufLowWaterMark) {
+                // insufficient space to allocate, due to alignment requirement
+                return NULL;
+            }
+        }
+    }
+
+    // result now points to a valid, aligned, unallocated memory block of appropriate size.
+    // update high/low water marks, and return the buffer.
+    if (allocateFromBottom) {
+        s_BigBufLowWaterMark = result + totalRequestedBytes;
+    } else {
+        s_BigBufHighWaterMark = result;
+    }
+
+    // finally, zero the memory before returning it.
+    memset(result, 0, totalRequestedBytes);
+    ++s_BigBufAllocationCount;
+    return result;
 }
 
+/// @brief Frees a block of memory that was allocated from the Big Buffer
+/// @param ptr Pointer to the memory block to be free'd.
+/// @note At least initially, the memory is not returned for re-use.
+/// @todo Allow the most-recent allocated memory to actually be free'd.
```

Comment:
This seems like the tricky bit (reuse). However - if this if primarily targeted at modes, then those structures all have a consistent lifecycle:
1. Enter mode
2. Allocate mode config struct
3. Exit mode
4. Release all memory

Allocating persistent ram for mode apps would have some extra complexities that I don't know how to handle. Does the app have a set of pointers that are null if unallocated, and !null when allocated? Those could also be cleared out when the mode exits.
6. Deallocate app ram from the top of the stack

Reply:
If all that's desired is support for what you listed above, then the logic(+) already in this PR.

Allocation of "persistent" RAM is also possible with very small changes to the logic:

* Default / temporary allocations come from bottom (128k aligned, to start)
* Long-term allocations come from top
* BigBuffer_Reset() would reset only to the current highwatermark (leaving any long-term allocations at the top part of the address space alone)

_**Please let me know what scenarios you'd like to enable!**_

(+) The logic is enough.  Requires testing / validation / and then fixing any bugs.

Reply:
Resolving as out-of-date with Alpha APIv2 commit